Applications/Spire/Source/KeyBindings/CancelKeyBindingsTableModel.cpp

```diff
@@ -0,0 +1,205 @@
+#include "Spire/KeyBindings/CancelKeyBindingsTableModel.hpp"
+
+using namespace Spire;
+using Action = CancelKeyBindingsTableModel::Action;
+using Binding = CancelKeyBindingsTableModel::Binding;
+
+namespace {
+  const auto ROW_COUNT = 13;
+  const auto COLUMN_COUNT = 2;
+
+  constexpr auto get_action(int index) {
+    switch(index) {
+      case 0:
+        return Action::ALL;
+      case 1:
+        return Action::ALL_ASKS;
+      case 2:
+        return Action::ALL_BIDS;
+      case 3:
+        return Action::CLOSEST_ASK;
+      case 4:
+        return Action::CLOSEST_BID;
+      case 5:
+        return Action::FURTHEST_BID;
+      case 6:
+        return Action::FURTHEST_ASK;
+      case 7:
+        return Action::MOST_RECENT;
+      case 8:
+        return Action::MOST_RECENT_ASK;
+      case 9:
+        return Action::MOST_RECENT_BID;
+      case 10:
+        return Action::OLDEST;
+      case 11:
+        return Action::OLDEST_ASK;
+      case 12:
+        return Action::OLDEST_BID;
+      default:
+        return Action::ALL;
+    }
+  }
+
+  constexpr auto get_index(Action action) {
+    switch(action) {
+      case get_action(0):
+        return 0;
+      case get_action(1):
+        return 1;
+      case get_action(2):
+        return 2;
+      case get_action(3):
+        return 3;
+      case get_action(4):
+        return 4;
+      case get_action(5):
+        return 5;
+      case get_action(6):
+        return 6;
+      case get_action(7):
+        return 7;
+      case get_action(8):
+        return 8;
+      case get_action(9):
+        return 9;
+      case get_action(10):
+        return 10;
+      case get_action(11):
+        return 11;
+      case get_action(12):
+        return 12;
+      default:
+        return -1;
+    }
+  }
+
+  constexpr auto get_action_text(int row) {
+    switch(row) {
+      case 0:
+        return std::string_view("All");
```

Comment:
```string_view``` is banned, return a ```const char*```, should not be a need for ```constexpr```.

Reply:
Does the return type imply constexpr?

Reply:
Nope, does this need to be ```constexpr```?

Reply:
I don't think it absolutely needs to be, but can't it be optimized if marked constexpr?

Reply:
It's a good thought for sure, but unfortunately ```constexpr``` is not an optimization and should not be used for optimization purposes. ```constexpr``` is used when you need to compute a value within a constant expression. For example you have ```switch``` statements with ```case``` clauses as follows:

    get_action(0)

That ```get_action``` must be a ```constexpr``` because each ```case``` requires a constant expression so there it is needed. But for optimizations do not use ```constexpr``` as it won't do you any good.